// @babylonjs/serializers/glTF/2.0/Extensions/EXT_lights_area.js
import { Vector3, Quaternion, TmpVectors } from "@babylonjs/core/Maths/math.vector.js";
import { Light } from "@babylonjs/core/Lights/light.js";
import { GLTFExporter } from "../glTFExporter.js";
import { Logger } from "@babylonjs/core/Misc/logger.js";
import { ConvertToRightHandedPosition, OmitDefaultValues, CollapseChildIntoParent, IsChildCollapsible } from "../glTFUtilities.js";
const NAME = "EXT_lights_area";
const DEFAULTS = {
    name: "",
    color: [1, 1, 1],
    intensity: 1,
    size: 1,
};
const RECTDEFAULTS = {
    aspect: 1,
};
const LIGHTDIRECTION = Vector3.Backward();
/**
 * [Specification](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/EXT_lights_area/README.md)
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
export class EXT_lights_area {
    /**
     * @internal
     */
    constructor(exporter) {
        /** The name of this extension. */
        this.name = NAME;
        /** Defines whether this extension is enabled. */
        this.enabled = true;
        /** Defines whether this extension is required */
        this.required = false;
        this._exporter = exporter;
    }
    /** @internal */
    dispose() {
        this._lights = null;
    }
    /** @internal */
    get wasUsed() {
        return !!this._lights;
    }
    /** @internal */
    onExporting() {
        this._exporter._glTF.extensions[NAME] = this._lights;
    }
    /**
     * Define this method to modify the default behavior when exporting a node
     * @param context The context when exporting the node
     * @param node glTF node
     * @param babylonNode BabylonJS node
     * @param nodeMap Node mapping of babylon node to glTF node index
     * @param convertToRightHanded Flag to convert the values to right-handed
     * @returns nullable INode promise
     */
    async postExportNodeAsync(context, node, babylonNode, nodeMap, convertToRightHanded) {
        return await new Promise((resolve) => {
            if (!(babylonNode instanceof Light)) {
                resolve(node);
                return;
            }
            const lightType = babylonNode.getTypeID() == Light.LIGHTTYPEID_RECT_AREALIGHT ? "rect" /* EXTLightsArea_LightType.RECT */ : null;
            if (!lightType) {
                Logger.Warn(`${context}: Light ${babylonNode.name} is not supported in ${NAME}`);
                resolve(node);
                return;
            }
            const areaLight = babylonNode;
            if (areaLight.falloffType !== Light.FALLOFF_GLTF) {
                Logger.Warn(`${context}: Light falloff for ${babylonNode.name} does not match the ${NAME} specification!`);
            }
            // Set the node's translation and rotation here, since lights are not handled in exportNodeAsync
            if (!areaLight.position.equalsToFloats(0, 0, 0)) {
                const translation = TmpVectors.Vector3[0].copyFrom(areaLight.position);
                if (convertToRightHanded) {
                    ConvertToRightHandedPosition(translation);
                }
                node.translation = translation.asArray();
            }
            // Represent the Babylon light's direction as a quaternion
            // relative to glTF lights' forward direction, (0, 0, -1).
            const direction = Vector3.Forward();
            if (convertToRightHanded) {
                ConvertToRightHandedPosition(direction);
            }
            const lightRotationQuaternion = Quaternion.FromUnitVectorsToRef(LIGHTDIRECTION, direction, TmpVectors.Quaternion[0]);
            if (!Quaternion.IsIdentity(lightRotationQuaternion)) {
                node.rotation = lightRotationQuaternion.asArray();
            }
            const light = {
                type: lightType,
                name: areaLight.name,
                color: areaLight.diffuse.asArray(),
                intensity: areaLight.intensity,
                size: areaLight.height,
                rect: {
                    aspect: areaLight.width / areaLight.height,
                },
            };
            OmitDefaultValues(light, DEFAULTS);
            if (light.rect) {
                OmitDefaultValues(light.rect, RECTDEFAULTS);
            }
            this._lights || (this._lights = {
                lights: [],
            });
            this._lights.lights.push(light);
            const lightReference = {
                light: this._lights.lights.length - 1,
            };
            // Assign the light to its parent node, if possible, to condense the glTF
            // Why and when: the glTF loader generates a new parent TransformNode for each light node, which we should undo on export
            const parentBabylonNode = babylonNode.parent;
            if (parentBabylonNode && IsChildCollapsible(areaLight, parentBabylonNode)) {
                const parentNodeIndex = nodeMap.get(parentBabylonNode);
                if (parentNodeIndex) {
                    // Combine the light's transformation with the parent's
                    const parentNode = this._exporter._nodes[parentNodeIndex];
                    CollapseChildIntoParent(node, parentNode);
                    parentNode.extensions || (parentNode.extensions = {});
                    parentNode.extensions[NAME] = lightReference;
                    // Do not export the original node
                    resolve(null);
                    return;
                }
            }
            node.extensions || (node.extensions = {});
            node.extensions[NAME] = lightReference;
            resolve(node);
        });
    }
}
GLTFExporter.RegisterExtension(NAME, (exporter) => new EXT_lights_area(exporter));
