// @babylonjs/viewer/dist/chunks/flowGraphSceneTickEventBlock-D8sBATJW.esm.js
import { c as FlowGraphEventBlock } from './KHR_interactivity-C-sP6iKo.esm.js';
import { g as RegisterClass } from './index-DLpEu_NE.esm.js';
import { f as RichTypeNumber } from './declarationMapper-mhjJHy6k.esm.js';
import './objectModelMapping-B8MY-rGq.esm.js';

/**
 * Block that triggers on scene tick (before each render).
 */
class FlowGraphSceneTickEventBlock extends FlowGraphEventBlock {
    constructor() {
        super();
        this.type = "SceneBeforeRender" /* FlowGraphEventType.SceneBeforeRender */;
        this.timeSinceStart = this.registerDataOutput("timeSinceStart", RichTypeNumber);
        this.deltaTime = this.registerDataOutput("deltaTime", RichTypeNumber);
    }
    /**
     * @internal
     */
    _preparePendingTasks(_context) {
        // no-op
    }
    /**
     * @internal
     */
    _executeEvent(context, payload) {
        this.timeSinceStart.setValue(payload.timeSinceStart, context);
        this.deltaTime.setValue(payload.deltaTime, context);
        this._execute(context);
        return true;
    }
    /**
     * @internal
     */
    _cancelPendingTasks(_context) {
        // no-op
    }
    /**
     * @returns class name of the block.
     */
    getClassName() {
        return "FlowGraphSceneTickEventBlock" /* FlowGraphBlockNames.SceneTickEvent */;
    }
}
RegisterClass("FlowGraphSceneTickEventBlock" /* FlowGraphBlockNames.SceneTickEvent */, FlowGraphSceneTickEventBlock);

export { FlowGraphSceneTickEventBlock };
