// @babylonjs/viewer/dist/chunks/flowGraphFunctionReferenceBlock-C7Tm2v83.esm.js
import { F as FlowGraphBlock } from './KHR_interactivity-C-sP6iKo.esm.js';
import { s as RichTypeString, R as RichTypeAny } from './declarationMapper-mhjJHy6k.esm.js';
import { g as RegisterClass } from './index-DLpEu_NE.esm.js';
import './objectModelMapping-B8MY-rGq.esm.js';

/**
 * A flow graph block that takes a function name, an object and an optional context as inputs and calls the function on the object.
 */
class FlowGraphFunctionReferenceBlock extends FlowGraphBlock {
    constructor(
    /**
     * the configuration of the block
     */
    config) {
        super(config);
        this.functionName = this.registerDataInput("functionName", RichTypeString);
        this.object = this.registerDataInput("object", RichTypeAny);
        this.context = this.registerDataInput("context", RichTypeAny, null);
        this.output = this.registerDataOutput("output", RichTypeAny);
    }
    _updateOutputs(context) {
        const functionName = this.functionName.getValue(context);
        const object = this.object.getValue(context);
        const contextValue = this.context.getValue(context);
        if (object && functionName) {
            const func = object[functionName];
            if (func && typeof func === "function") {
                this.output.setValue(func.bind(contextValue), context);
            }
        }
    }
    getClassName() {
        return "FlowGraphFunctionReference" /* FlowGraphBlockNames.FunctionReference */;
    }
}
RegisterClass("FlowGraphFunctionReference" /* FlowGraphBlockNames.FunctionReference */, FlowGraphFunctionReferenceBlock);

export { FlowGraphFunctionReferenceBlock };
