// @babylonjs/viewer/dist/chunks/flowGraphArrayIndexBlock-Dz0W6BC1.esm.js
import { F as FlowGraphBlock, g as getNumericValue } from './KHR_interactivity-C-sP6iKo.esm.js';
import { R as RichTypeAny, b as FlowGraphInteger } from './declarationMapper-mhjJHy6k.esm.js';
import { g as RegisterClass } from './index-DLpEu_NE.esm.js';
import './objectModelMapping-B8MY-rGq.esm.js';

/**
 * This simple Util block takes an array as input and selects a single element from it.
 */
class FlowGraphArrayIndexBlock extends FlowGraphBlock {
    /**
     * Construct a FlowGraphArrayIndexBlock.
     * @param config construction parameters
     */
    constructor(config) {
        super(config);
        this.config = config;
        this.array = this.registerDataInput("array", RichTypeAny);
        this.index = this.registerDataInput("index", RichTypeAny, new FlowGraphInteger(-1));
        this.value = this.registerDataOutput("value", RichTypeAny);
    }
    /**
     * @internal
     */
    _updateOutputs(context) {
        const array = this.array.getValue(context);
        const index = getNumericValue(this.index.getValue(context));
        if (array && index >= 0 && index < array.length) {
            this.value.setValue(array[index], context);
        }
        else {
            this.value.setValue(null, context);
        }
    }
    /**
     * Serializes this block
     * @param serializationObject the object to serialize to
     */
    serialize(serializationObject) {
        super.serialize(serializationObject);
    }
    getClassName() {
        return "FlowGraphArrayIndexBlock" /* FlowGraphBlockNames.ArrayIndex */;
    }
}
RegisterClass("FlowGraphArrayIndexBlock" /* FlowGraphBlockNames.ArrayIndex */, FlowGraphArrayIndexBlock);

export { FlowGraphArrayIndexBlock };
