// @babylonjs/viewer/dist/chunks/flowGraphBranchBlock-Bc01R8yg.esm.js
import { h as RichTypeBoolean } from './declarationMapper-mhjJHy6k.esm.js';
import { k as FlowGraphExecutionBlock } from './KHR_interactivity-C-sP6iKo.esm.js';
import { g as RegisterClass } from './index-DLpEu_NE.esm.js';
import './objectModelMapping-B8MY-rGq.esm.js';

/**
 * A block that evaluates a condition and activates one of two branches.
 */
class FlowGraphBranchBlock extends FlowGraphExecutionBlock {
    constructor(config) {
        super(config);
        this.condition = this.registerDataInput("condition", RichTypeBoolean);
        this.onTrue = this._registerSignalOutput("onTrue");
        this.onFalse = this._registerSignalOutput("onFalse");
    }
    _execute(context) {
        if (this.condition.getValue(context)) {
            this.onTrue._activateSignal(context);
        }
        else {
            this.onFalse._activateSignal(context);
        }
    }
    /**
     * @returns class name of the block.
     */
    getClassName() {
        return "FlowGraphBranchBlock" /* FlowGraphBlockNames.Branch */;
    }
}
RegisterClass("FlowGraphBranchBlock" /* FlowGraphBlockNames.Branch */, FlowGraphBranchBlock);

export { FlowGraphBranchBlock };
